(PDB-2640) Add option for gzip compressing a request's body

camlow325 opened the pull request:
This commit adds a new Clojure HTTP request option,
:compress-request-body`, and corresponding Java option which can be used
to have gzip compression applied to the request's body before it is sent
along to the server.

puppetcla commented:
CLA signed by all contributors.

jpinsonault commented:
Looks like it needs a rebase

camlow325 commented on src/java/com/puppetlabs/http/client/Sync.java:
```diff
@@ -31,7 +31,12 @@ private static RequestOptions extractRequestOptions(SimpleRequestOptions simpleO
+                simpleOptions.getCompressRequestBody();
+        RequestOptions requestOptions = new RequestOptions(
+                uri, headers, body, decompressBody, as);
+        requestOptions.setCompressRequestBody(requestBodyDecompression);
```
`RequestOptions` is public API at this point so I don't think I could safely just change the existing constructor to take an additional argument without potentially breaking an existing client.  I could add an overloaded constructor which takes the extra argument.  Since I'm not sure how often this option would be used by consumers outside of the one I'm thinking of today, the PuppetDB terminus, just having a way to access it via a setter without creating another constructor seemed like a better way to go for now.  Does that make sense?

jpinsonault replied:
Yeah that makes sense, this seems sane then

camlow325 commented on src/java/com/puppetlabs/http/client/impl/JavaClient.java:
```diff
@@ -80,6 +94,10 @@
+        if (options.getCompressRequestBody() == CompressType.GZIP &&
+                (! result.containsKey("content-encoding"))) {
+            result.put("content-encoding", new BasicHeader("Content-Encoding", "gzip"));
+        }
```
Yeah, it might cause a problem.  Depends upon whether the client and server might be doing something "custom" which would allow the server to still end up decoding the request body properly.  I figured it would be safer to follow the pattern we used previously for other headers - like with `accept-encoding` above - where if the client specified a value for the header that we should just trust that the client wants that value to go out verbatim in the request rather than overriding the client's intent.  Does that seem reasonable?

jpinsonault replied:
Ah ok, if that's how this is used then it seems good to defer to the value that's already there

camlow325 commented:
Thanks, @jpinsonault.  I rebased this at 689db7c.  The only conflict I had to resolve was around the `cheshire` dev dependency I had added, which conflicted with the upgrade to clj-parent dependencies done in #68.

camlow325 commented:
@rlinehan or @scottyw - could at least one of you take a look at this one?  Thanks!

scottyw commented:
I spent many a happy hour with that trying to make that PipedXxxStream stuff work in unbuffered responses. It's a nightmare. What you've done looks pretty fine to me.

camlow325 commented:
@scottyw Thanks very much for reviewing my code!  Agree that the PipedXxxStream stuff is a nightmare to follow.  😏 

I found one hole in my code related to how the input stream is closed when a gzip-compressed request has been completed.  I addressed this in cb5de45.  If someone could take a look at that, would be much appreciated.  Thanks!
